domain/application/service/application.go

```diff
@@ -224,132 +248,214 @@ func (s *ApplicationService) CreateApplication(
 	args AddApplicationArgs,
 	units ...AddUnitArg,
 ) (coreapplication.ID, error) {
+	modelType, appArg, err := s.makeCreateApplicationArgs(ctx, name, charm, origin, args)
```

Comment:
This is to avoid excess db calls. The model type is looked up and actually used in the `makeCreateApplicationArgs` method. It's also needed below inside a for loop. So given it's looked up once, it is returned and used below. This is an internal service method. What implementation are we possible bleeding out? Internal methods by definition share implementation knowledge.

Reply:
Ok, I've split the validation and arg creation into separate methods to avoid getting the model type and then not using it.